File: Queues/649. Dota2 Senate/649. Dota2 Senate.py
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        r_queue = deque()
        d_queue = deque()

        for i, s in enumerate(senate):
            if s == "R":
                r_queue.append(i)
            else:
                d_queue.append(i)

        while (r_queue and d_queue):
            r = r_queue.popleft()
            d = d_queue.popleft()
            if r < d:
                r_queue.append(r + len(senate))
            else:
                d_queue.append(d + len(senate))
        
        if (r_queue):
            return "Radiant"
        return "Dire"

# Time: O(n)
# Space: O(n)

File: Queues/649. Dota2 Senate/649. Dota2 Senate.js
/**
 * @param {string} senate
 * @return {string}
 */
var predictPartyVictory = function (senate) {
  let r_queue = [];
  let d_queue = [];

  for (let i = 0; i < senate.length; i++) {
    if (senate[i] == "R") r_queue.push(i);
    else d_queue.push(i);
  }

  while (r_queue.length > 0 && d_queue.length > 0) {
    const r = r_queue.shift();
    const d = d_queue.shift();
    if (r < d) r_queue.push(r + senate.length);
    else d_queue.push(d + senate.length);
  }

  if (r_queue.length > 0) return "Radiant";
  return "Dire";
};

// Time: O(n)
// Space: O(n)
